File: src/main/java/nz/co/searchwellington/controllers/models/ModelBuilder.java
package nz.co.searchwellington.controllers.models;

import javax.servlet.http.HttpServletRequest;

import org.springframework.web.servlet.ModelAndView;

public interface ModelBuilder {

	public abstract boolean isValid(HttpServletRequest request);

	public abstract ModelAndView populateContentModel(HttpServletRequest request);
	
	public void populateExtraModelContent(HttpServletRequest request, ModelAndView mv);
	
	public String getViewName(ModelAndView mv);

}

File: src/main/scala/nz/co/searchwellington/controllers/models/helpers/ModelBuilder.scala
package nz.co.searchwellington.controllers.models.helpers

import io.opentelemetry.api.trace.Span
import jakarta.servlet.http.HttpServletRequest
import nz.co.searchwellington.controllers.CommonModelObjects
import nz.co.searchwellington.model.User
import nz.co.searchwellington.repositories.ContentRetrievalService
import org.springframework.ui.ModelMap

import scala.concurrent.{ExecutionContext, Future}

trait ModelBuilder extends ContentFields with CommonModelObjects {

  def contentRetrievalService: ContentRetrievalService

  def isValid(request: HttpServletRequest): Boolean

  def populateContentModel(request: HttpServletRequest, loggedInUser: Option[User] = None)(implicit ec: ExecutionContext, currentSpan: Span): Future[Option[ModelMap]]

  def populateExtraModelContent(request: HttpServletRequest, loggedInUser: Option[User])(implicit ec: ExecutionContext, currentSpan: Span): Future[ModelMap]

  def getViewName(mv: ModelMap, loggedInUser: Option[User]): String

}
